docs/proposals/operator-sdk-v1-migration.md

```diff
@@ -0,0 +1,285 @@
+   but is not necessary if we do not use `openapi-gen`.
+
+We currently use `deepcopy-gen` markers to generate our deep copy code, and use
+`k8s.io/code-generator` to create our OpenAPI generated code under the covers. Presumably using
+the `kubebuilder` marker will let us retain our generated deep copies.
+
+[4] https://sdk.operatorframework.io/docs/building-operators/golang/migration/#apis
+
+### Controllers
+
+Controller code can more or less be migrated as is from
+`pkg/controller/<kind>/<kind>_controller.go` to `/controllers/<kind>_controller.go` [5]. There are
+a few subtleties with generated types and a change to the manager interface, but in general the
+shift is straightforward.
+
+[5] https://sdk.operatorframework.io/docs/building-operators/golang/migration/#controllers
+
+### RBAC
+
+RBAC is maintained via kubebuilder RBAC Markers, rather than separate YAML files. The RBAC markers
+are placed on the `Reconcile` method of each controller, alongside implementing code. SDK projects
+can then generate RBAC for the controllers via a make target (`make manifests`). RBAC YAML has been
+moved to `config/rbac/` directory, rather than living alongside deployment manifests in `deploy`.
+
+New operators are cluster-scoped. We believe this is the right scoping for the build controllers.
+
+[6] https://sdk.operatorframework.io/docs/building-operators/golang/migration/#set-the-rbac-permissions
+[7] https://sdk.operatorframework.io/docs/building-operators/golang/operator-scope/
+
+### Migrate main.go
+
+The main invocation has changed its default generated code with respect to leader election.
+Operator SDK has not deprecated `leader.Become` (yet), but it has moved new code to use the leader
+election libraries from controller-runtime [8].
+
+If we maintain usage of `leader.Become`, there are changes to default behavior which we need to
+account for:
+
+1. The default metrics endpoint now binds to port `:8080` instead of `:8383`. To use the old port,
+   we need to pass the `--metrics-addr` argument to our binary.
+2. `OPERATOR_NAME` and `POD_NAME` env vars are no longer used for leader election. Instead, a
+   leader election ID is provided to the controller manager on startup. We may be ignoring these
+   environment variables already by hard-coding the LeaderElectionID field when the controller
+   manager is started up (`build-operator-lock`).
+
+Our operator allows the leader election namespace to be customized via the
+`BUILD_OPERATOR_LEADER_ELECTION_NAMESPACE` environment variable. This is ultimately wired through
+to controller-runtime's leader election logic - this should remain stable post migration.
+
+In addition to the old metrics port, our current controller also exposes port `:8686` to serve the
+custom resource metrics. Altering the metrics port could be considered a breaking change, as we
+would alter the related Kubernetes service. It is unclear if we can continue this practice.
```

Comment:
I would go for the breaking change, provided we do it soon enough.

Reply:
@zhangtbj do u recall why the `8686` here?

Reply:
I remember `8686` is the operatorMetricsPort, it is the controller port to serve the controller related metrics like queue, go_alloc, etc...., at the beiginning, I found the custom metrics which we defined can also be forwarded to this port directly. So we use this port directly too.